### PR TITLE
added middlewares for CORS and no-cache headers to grpc-gateway

### PR DESCRIPTION
**Pull request by Ekliptor**
This also fixes the current issue that nginx CORS headers are only sent on HTTP 200 responses but not on errors (which is why we couldn't parse error messages in browser).

Middlewares run always before the gateway see comment:
```
	// Apply middlewares to modify all responses. Our middleware run from inner to outer function
	// Example: Func2(Func1(mux))
	// To make a middleware run first, you must inside call next.ServeHTTP() before modifying the response.
	return http.ListenAndServe(":"+*proxyPort, middlewares.CorsMiddleware(middlewares.NoCacheMiddleware(mux)))
```

### CODE REVIEW

File: bchrpc/proxy/gw.go [hunk 1]
```diff
@@ -46,7 +47,10 @@ func run() error {
 	// TODO: consider serving static files for swagger-ui
 
 	// Start HTTP server (and proxy calls to gRPC server endpoint)
-	return http.ListenAndServe(":"+*proxyPort, mux)
+	// Apply middlewares to modify all responses. Our middleware run from inner to outer function
+	// Example: Func2(Func1(mux))
+	// To make a middleware run first, you must inside call next.ServeHTTP() before modifying the response.
+	return http.ListenAndServe(":"+*proxyPort, middlewares.CorsMiddleware(middlewares.NoCacheMiddleware(mux)))
```
Comment (jcramer) [inline bchrpc/proxy/gw.go:53]: Can we refactor this to be like:

```
// Apply middlewares to modify all responses.
middlewares.NoCacheMiddleware(mux)
middlewares.CorsMiddleware(mux)

// Start HTTP server (and proxy calls to gRPC server endpoint)
return http.ListenAndServe(":"+*proxyPort, mux)
```

File: bchrpc/proxy/middlewares/cors.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"net/http"
+)
+
+// Overrides CORS headers per request.
+func CorsMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// programatically check the origin against a map of allowed origins
+		/*
```
Comment (jcramer) [inline bchrpc/proxy/middlewares/cors.go:11]: Can we remove this commented code block?

**Review by jcramer — CHANGES_REQUESTED** [review]
Looks good, see a couple of minor comments.

File: bchrpc/proxy/middlewares/cors.go (hunk 2 above)
Comment (Ekliptor) [reply bchrpc/proxy/middlewares/cors.go]: I removed the code block. our strange runtime router doesn't allow applying middlewares like this.

**Review by jcramer — APPROVED** [review]

**Pull request merged**